From ea9a2d7a9bdf7977b051f8ab8160fbfc224d2636 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 30 Sep 2025 17:53:38 +0200
Subject: [PATCH 924/954] feat(fdt-wrappers): fdt_get_phandle_with_args()
 function

This function parses a phandle with arguments from a device tree
node property.

Example:
node {
    clocks = <&rcc 1>;
};

Usage:
err = fdt_get_phandle_with_args(fdt, node, "clocks", "#clock-cells",
                                0, &args);

Result:
  args.args_count = 1
  args.phandle = phandle of RCC
  args.args[0] = 1

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 6d7973dc9c229f4482029028e9705b8bcd4381f6 feat(fdt-wrappers): fdt_get_phandle_with_args() function]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I30477bf73f2ab23ffa580b7ad7baa3bc96cafb9f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/495177
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 common/fdt_wrappers.c         | 46 +++++++++++++++++++++++++++++++++++
 include/common/fdt_wrappers.h | 12 +++++++++
 2 files changed, 58 insertions(+)

diff --git a/common/fdt_wrappers.c b/common/fdt_wrappers.c
index 9138716d14..6609e18632 100644
--- a/common/fdt_wrappers.c
+++ b/common/fdt_wrappers.c
@@ -652,3 +652,49 @@ int fdtw_find_or_add_subnode(void *fdt, int parentoffset, const char *name)
 
 	return offset;
 }
+
+/* Parse a phandle with arguments from a property of a device tree node */
+int fdt_get_phandle_with_args(void *fdt, int node, const char *list_name,
+			      const char *cells_name, int index,
+			      struct fdt_phandle_args *out_args)
+{
+	const fdt32_t *cuint;
+	uint32_t phandle;
+	uint32_t cells;
+	int entry_size;
+	int prop_len;
+	int i = 0;
+
+	cuint = fdt_getprop(fdt, node, list_name, &prop_len);
+	if (!cuint || prop_len <= 0) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	while (prop_len > 0) {
+
+		phandle = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(*cuint));
+
+		if (fdt_read_uint32(fdt, phandle, cells_name, &cells) < 0) {
+			return -FDT_ERR_NOTFOUND;
+		}
+
+		entry_size = (1 + cells) * sizeof(fdt32_t);
+		if (prop_len < entry_size) {
+			return -FDT_ERR_BADVALUE;
+		}
+
+		if (i == index) {
+			out_args->phandle = phandle;
+			out_args->args_count = cells;
+			out_args->args[0] = (unsigned long)fdt32_to_cpu(*(cuint + 1));
+
+			return 0;
+		}
+
+		cuint += 1 + cells;
+		prop_len -= entry_size;
+		i++;
+	}
+
+	return -FDT_ERR_NOTFOUND;
+}
diff --git a/include/common/fdt_wrappers.h b/include/common/fdt_wrappers.h
index 9a5972859c..05eb2834ac 100644
--- a/include/common/fdt_wrappers.h
+++ b/include/common/fdt_wrappers.h
@@ -15,6 +15,14 @@
 /* Number of cells, given total length in bytes. Each cell is 4 bytes long */
 #define NCELLS(len) ((len) / 4U)
 
+#define MAX_PHANDLE_ARGS 16
+
+struct fdt_phandle_args {
+	uint32_t phandle;
+	int args_count;
+	uint32_t args[MAX_PHANDLE_ARGS];
+};
+
 int fdt_read_uint32(const void *dtb, int node, const char *prop_name,
 		    uint32_t *value);
 uint32_t fdt_read_uint32_default(const void *dtb, int node,
@@ -71,4 +79,8 @@ for (node = fdt_node_offset_by_compatible(dtb, -1, compatible_str);   \
      node >= 0;                                                       \
      node = fdt_node_offset_by_compatible(dtb, node, compatible_str))
 
+int fdt_get_phandle_with_args(void *fdt, int node, const char *list_name,
+			      const char *cells_name, int index,
+			      struct fdt_phandle_args *out_args);
+
 #endif /* FDT_WRAPPERS_H */
-- 
2.53.0

